org.adempiere.base/src/org/compiere/model/MInvoice.java

```diff
@@ -3347,4 +3351,70 @@ public void updateFrom(MOrder order)
 			}
 		}
 	}
+	
+	/**
+	 * Allocate Invoice for Completed Unallocated Pre-payments
+	 */
+	private boolean allocateFromOrder() {
+		// Get Set Of OrderIds from Lines
+		HashSet<Integer> orderIDSet = new HashSet<>();
+
+		if(getC_Order_ID() > 0)
+			orderIDSet.add(getC_Order_ID());
+
+		MInvoiceLine[] lines = getLines();
+		for(MInvoiceLine line : lines) {
+			if(line.getC_OrderLine_ID() <= 0)
+				continue;
+
+			 I_C_OrderLine orderLine = line.getC_OrderLine();
+			 if(orderLine == null)
+				 continue;
+
+			 orderIDSet.add(orderLine.getC_Order_ID());
+		}
+
+		if(orderIDSet.size() <= 0)
+			return false;
+
+		String whereClause = "docstatus = 'CO' AND NOT EXISTS( SELECT 1 FROM c_allocationline al where al.c_payment_id=c_payment.c_payment_id) AND IsAllocated='N' ";

```

Comment:
since there's need to append to the string, it is more efficient to use stringbuilder here instead.